kernel/of/fdt_mem.c

```diff
@@ -0,0 +1,192 @@
+
+#include <kernel/of/fdt_mem.h>
+#include <stdio.h>
+
+// Helper function which decides how to deal with overlapping reserved regions
+// Returns true if split, else false
+static bool
+trim_or_split_around_reserved(
+    void** start_ptr,
+    size_t* size_ptr, // Original Region to be trimmed or split
+    void** split_start_ptr,
+    size_t* split_size_ptr, // Storage for second region if split occurs
+    void* resv_start,
+    size_t resv_size // The reserved region to trim or split around
+)
+{
+    void* start = *start_ptr;
+    size_t size = *size_ptr;
+
+    void* end = start + size;
+    void* resv_end = resv_start + resv_size;
+
+    if (resv_end <= start || resv_start > end) {
+        // No overlap with this reserve entry
+        return false;
+    }
+    else if (resv_start <= start) {
+        // Fully covering or trim from below
+        if (resv_end >= end) {
+            // Fully covering (this region is entirely reserved)
+            *start_ptr = NULL;
+            *size_ptr = 0;
+            return false;
+        }
+        else {
+            // Trim from below
+            *start_ptr = resv_end;
+            *size_ptr = (size_t)(end - resv_end);
+            return false;
+        }
+    }
+    else {
+        // Split into two or trim from above
+        if (resv_end < end) {
+            // Split into two
+            // Region1 = [start,resv_start)
+            // Region2 = [resv_end,end)
+            *size_ptr = (size_t)(resv_start - start);
+            *split_start_ptr = resv_end;
+            *split_size_ptr = (size_t)(end - *split_start_ptr);
+            return true;
+        }
+        else {
+            // Trim from above
+            end = resv_start;
+            *size_ptr = (size_t)(end - start);
+            return false;
+        }
+    }
+
+    // Should be unreachable
+}
+
+// Try to free this region (return the number of bytes successfully freed)
+static size_t
+try_boot_free_region(struct fdt* fdt, void* start, size_t size)
+{
+    // Should be unnecessary (but it's worth checking in case someone messes up)
+    if (size == 0) {
+        return (size_t)0;
+    }
+
+    // Buffers for a second region if we get split into two
+    void* split_start;
+    size_t split_size;
+
+    // Check if this is a region before the end of the kernel (protects the kernel and
+    // firmware which is before it)
+    extern int __kernel_end[];
```

Comment:
```suggestion
    extern void* __kernel_end;
```

this isn't defined in our linker script though

Reply:
also i think you have to take the address of it, like I did with `__init_array_start` in `arch/riscv/kernel.c`

Reply:
It is defined in the linker script, and it's defined as an array to avoid taking the address. It's of type `int` to avoid people  who don't know how linker script variables work from thinking the "value" of the `void*` is the end of the kernel instead of the symbol itself

Reply:
This way we also are correct even if people mistakenly try to take `&__kernel_end` because `__kernel_end == &__kernel_end` for array types.

Reply:
Alright sure but shouldn't it be declared `uintptr_t` then?